File: CPP/No676.cc
/**
 * Created by Xiaozhong on 2020/11/13.
 * Copyright (c) 2020/11/13 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>
#include <unordered_map>
#include <unordered_set>

using namespace std;

class MagicDictionary {
private:
    unordered_map<string, int> count;
    unordered_set<string> words;

    vector<string> generalizedNeighbors(string &word) {
        vector<string> ret;
        for (int i = 0; i < word.size(); ++i) {
            char letter = word[i];
            word[i] = '*';
            ret.emplace_back(word);
            word[i] = letter;
        }
        return ret;
    }

public:
    /** Initialize your data structure here. */
    MagicDictionary() {}

    void buildDict(vector<string> dictionary) {
        for (string word : dictionary) {
            words.insert(word);
            for (string &neighboor: generalizedNeighbors(word)) {
                if (count.count(neighboor)) {
                    count[neighboor] = count[neighboor] + 1;
                } else {
                    count.insert(make_pair(neighboor, 1));
                }
            }
        }
    }

    bool search(string searchWord) {
        for (string &neighboor:generalizedNeighbors(searchWord)) {
            int c = count[neighboor];
            if (c > 1 || c == 1 && !words.count(searchWord)) return true;
        }
        return false;
    }
};

int main() {
    vector<string> dictionary = {"hello", "hallo", "leetcode"};
    MagicDictionary md;
    md.buildDict(dictionary);
    cout << md.search("hhllo") << endl;
    cout << md.search("hello") << endl;
    cout << md.search("hallo") << endl;
}

File: Java/src/No676.java
/**
 * Java
 *
 * @Description: TODO
 * @Author: cherry
 * @Create on: 2022/7/11
 **/

public class No676 {
    private static class MagicDictionary {

        public MagicDictionary() {

        }

        public void buildDict(String[] dictionary) {
            for (String word : dictionary) insert(word);
        }

        public boolean search(String searchWord) {
            return dfs(root, searchWord, 0, false);
        }

        private boolean dfs(Node curr, String word, int index, boolean flag) {
            if (index == word.length()) return flag && curr.isEnd;
            if (curr.children[word.charAt(index) - 'a'] != null) {
                if (dfs(curr.children[word.charAt(index) - 'a'], word, index + 1, flag)) return true;
            }
            if (!flag) {
                for (int idx = 0; idx < 26; ++idx) {
                    if (idx != (word.charAt(index) - 'a') && curr.children[idx] != null) {
                        if (dfs(curr.children[idx], word, index + 1, true)) return true;
                    }
                }
            }
            return false;
        }

        private void insert(String word) {
            Node curr = root;
            for (char ch : word.toCharArray()) {
                if (curr.children[ch - 'a'] == null) curr.children[ch - 'a'] = new Node();
                curr = curr.children[ch - 'a'];
            }
            curr.isEnd = true;
        }

        private Node root = new Node();

        private static class Node {
            boolean isEnd = false;
            Node[] children = new Node[26];
        }
    }

    public static void main(String[] args) {
        MagicDictionary dictionary = new MagicDictionary();
        dictionary.buildDict(new String[]{"hello", "hallo", "leetcode"});
        System.out.println(dictionary.search("hhllo"));
    }
}
